Mathlib/GroupTheory/DivisibleHull.lean

```diff
@@ -0,0 +1,291 @@
+    mk m s = mk m' s' ↔ s' • m = s • m' := by
+  rw [mk_eq_mk]
+  constructor
+  · intro ⟨u, h⟩
+    simp_rw [Submonoid.smul_def] at ⊢ h
+    exact (smul_right_inj (ne_of_gt u.prop)).mp h
+  · intro hr
+    use 1
+    rw [hr]
+
+theorem mk_add_mk {m1 m2 : M} {s1 s2 : Submonoid.pos ℤ} :
+    mk m1 s1 + mk m2 s2 = mk (s2 • m1 + s1 • m2) (s1 * s2) := LocalizedModule.mk_add_mk
+
+theorem qsmul_mk (a : ℚ) (m : M) (s : Submonoid.pos ℤ) :
+    a • mk m s = mk (a.num • m) (⟨a.den, Int.natCast_pos.mpr a.den_pos⟩ * s) := by
+  convert LocalizedModule.mk'_smul_mk ℚ a.num m ⟨a.den, Int.natCast_pos.mpr a.den_pos⟩ s
+  rw [IsLocalization.eq_mk'_iff_mul_eq]
+  simp
+
+theorem zsmul_mk (a : ℤ) (m : M) (s : Submonoid.pos ℤ) : a • mk m s = mk (a • m) s := by
+  suffices (a : ℚ) • mk m s = mk (a • m) s by
+    rw [Int.cast_smul_eq_zsmul] at this
+    exact this
+  rw [qsmul_mk]
+  exact congrArg _ <| Subtype.eq (by simp)
+
+theorem nsmul_mk (a : ℕ) (m : M) (s : Submonoid.pos ℤ) : a • mk m s = mk (a • m) s := by
+  suffices (a : ℤ) • mk m s = mk ((a : ℤ) • m) s by simpa using this
+  exact zsmul_mk _ _ _
+
+instance : DivisibleBy (DivisibleHull M) ℕ where
+  div m n :=
+    if h : n = 0 then
+      0
+    else
+      liftOn m (fun m s ↦ mk m ⟨s.val * n,
+        Submonoid.mem_pos.mpr <| mul_pos s.prop (by simpa using Nat.pos_of_ne_zero h)⟩)
+      (fun m m' s s' ↦ by
+        suffices ∀ a : ℤ, 0 < a → a • s' • m = a • s • m' →
+          ∃ a : ℤ, 0 < a ∧ a • (s'.val * n) • m = a • (s.val * n) • m' by simpa [mk_eq_mk]
+        refine fun a ha h ↦ ⟨a, ha, ?_⟩
+        simp_rw [Submonoid.smul_def] at h
+        simp_rw [smul_smul, ← mul_assoc, mul_comm _ (n : ℤ)]
+        simp_rw [mul_smul]
+        rw [h])
+  div_zero := by simp
+  div_cancel n {m} h := by
+    induction m using ind with | mk m s
+    suffices ∃ a : ℤ, 0 < a ∧ a • s • n • m = a • (s.val * n) • m by simpa [h, nsmul_mk, mk_eq_mk]
+    use 1
+    suffices s • n • m = (s.val * n) • m by simpa
+    rw [mul_smul, natCast_zsmul, Submonoid.smul_def]
+
+variable [LinearOrder M] [IsOrderedAddMonoid M]
+
+instance : LE (DivisibleHull M) where
+  le x y := liftOn₂ x y (fun m s n t ↦ t • m ≤ s • n) fun m n m' n' s t s' t' h h' ↦ by
+    rw [mk_eq_mk_iff_smul_eq_smul] at h h'
+    suffices t • m ≤ s • n ↔ t' • m' ≤ s' • n' by simpa
+    simp_rw [Submonoid.smul_def] at ⊢ h h'
+    rw [← zsmul_le_zsmul_iff_right (mul_pos s'.prop t'.prop)]
+    rw [show (s'.val * t'.val) • t.val • m = (t.val * t'.val) • s'.val • m by
+      simp_rw [smul_smul]
+      ring_nf]
+    rw [show (s'.val * t'.val) • s.val • n = (s.val * s'.val) • t'.val • n by
+      simp_rw [smul_smul]
+      ring_nf]
+    rw [h, h']
+    rw [show (t.val * t'.val) • s.val • m' = (s.val * t.val) • t'.val • m' by
+      simp_rw [smul_smul]
+      ring_nf]
+    rw [show (s.val * s'.val) • t.val • n' = (s.val * t.val) • s'.val • n' by
+      simp_rw [smul_smul]
+      ring_nf]
+    exact zsmul_le_zsmul_iff_right (mul_pos s.prop t.prop)
```

Comment:
Again untested, but could you use `convert zsmul_le_zsmul_iff_right (mul_pos s.prop t.prop)` before that long chain of `rw`s? Hopefully that would then give you the expressions inside your `show` as goals.